### PR TITLE
[GlobalISel] Add G_ABDS and G_ABDU instructions

### PR DESCRIPTION
**Pull request by tschuett**
The DAG has the same instructions: the signed and unsigned absolute difference of it's input. For AArch64, they map to uabd and sabd for Neon and SVE. The Neon and SVE instructions will require custom patterns.

They are pseudo opcodes and are not imported by the IRTranslator. We need combines to create them.

PowerPC, ARM, and AArch64 have native instructions.

/// i.e trunc(abs(sext(Op0) - sext(Op1))) becomes abds(Op0, Op1) 
///  or trunc(abs(zext(Op0) - zext(Op1))) becomes abdu(Op0, Op1)

For GlobalISel, we are going to write the combines in MIR patterns.

see:
llvm/test/CodeGen/AArch64/abd-combine.ll

- [ ] combine into abd
- [ ] legalize and add td patterns

### CODE REVIEW

**Comment by tschuett** [comment]
The DAG combine:
https://github.com/llvm/llvm-project/blob/be75a14207ab0dac6ad9f9097957f14a918ac611/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L11081

**Comment by tschuett** [comment]
```
def combine_abds : GICombineRule<
  (defs root:$root, build_fn_matchinfo:$matchinfo),
  (match (G_SEXT $lhs, $x),
         (G_SEXT $rhs, $y),
         (G_SUB $sub, $lhs, $rhs),
         (G_ABS $abs, $sub),
         (G_TRUNC $root, $abs):$trunc,
         [{ return Helper.matchAbds(*${trunc}, ${matchinfo}); }]),
  (apply [{ Helper.applyBuildFn(*${trunc}, ${matchinfo}); }])>;
```

**Comment by tschuett** [comment]
X86 lowers them sometimes into selects. We could add the reverse combine:
// abds(lhs, rhs) -> select(slt(lhs,rhs),sub(rhs,lhs),sub(lhs,rhs))
// abdu(lhs, rhs) -> select(ult(lhs,rhs),sub(rhs,lhs),sub(lhs,rhs))

**Comment by tschuett** [comment]
RISCV lowering:
// abds(lhs, rhs) -> sub(smax(lhs,rhs), smin(lhs,rhs))
// abdu(lhs, rhs) -> sub(umax(lhs,rhs), umin(lhs,rhs))

**Comment by tschuett** [comment]
Alternative / teaser:
Add 20 pseudo opcodes, map them to AArch64ISDs (AArch64usdot), and add a ton of combines into the pseudos.

File: llvm/lib/CodeGen/MachineVerifier.cpp [hunk 1]
```diff
@@ -1585,6 +1585,41 @@ void MachineVerifier::verifyPreISelGenericInstruction(const MachineInstr *MI) {
...
+
+    if (SrcTy.isPointerOrPointerVector() || SrcTy2.isPointerOrPointerVector()) {
+      report("Generic abds/abdu does not support pointers as operands", MI);
+      break;
+    }
+
+    if (DstTy.isPointerOrPointerVector()) {
+      report("Generic abds/abdu does not support pointers as a result", MI);
+      break;
+    }
```
Comment (arsenm) [inline llvm/lib/CodeGen/MachineVerifier.cpp:1602]: Does this not get verified by default? I thought there were separate type and ptype to distinguish them

File: llvm/test/MachineVerifier/test_abd_su.mir [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    %15:_(s32) = G_CONSTANT i32 0
+    %16:_(s64) = G_CONSTANT i64 2
+    ; CHECK: Generic abds/abdu must have same input types
+    %17:_(s1) = G_ABDU %15, %16
+
+    %18:_(s64) = G_CONSTANT i64 0
+    %19:_(s64) = G_CONSTANT i64 2
+    ; CHECK: Generic abds/abdu must have same input and output types
+    %20:_(s1) = G_ABDU %18, %19
+
```
Comment (arsenm) [inline llvm/test/MachineVerifier/test_abd_su.mir:33]: Maybe check vector of pointer since that manages to always break

**Review by arsenm — APPROVED** [review]

File: llvm/lib/CodeGen/MachineVerifier.cpp (hunk 1 above)
Comment (tschuett) [reply llvm/lib/CodeGen/MachineVerifier.cpp:1602]: Good point. It was copy-pasted/inspired by the G_SCMP and G_USCMP nearby.

File: llvm/test/MachineVerifier/test_abd_su.mir [hunk 3]
```diff
@@ -6,16 +6,6 @@ name:            g_abd_su
 body: |
   bb.0:
 
-    %2:_(p0) = G_IMPLICIT_DEF
-    %3:_(p0) = G_IMPLICIT_DEF
-    ; CHECK: Generic abds/abdu does not support pointers as operands
```
Comment (arsenm) [inline llvm/test/MachineVerifier/test_abd_su.mir]: You would still keep these tests even if the generic logic handles it

File: llvm/test/MachineVerifier/test_abd_su.mir [hunk 4]
```diff
@@ -6,6 +6,14 @@ name:            g_abd_su
 body: |
   bb.0:
 
+    %2:_(p0) = G_IMPLICIT_DEF
+    %3:_(p0) = G_IMPLICIT_DEF
+    %4:_(s1) = G_ABDS %2, %3
```
Comment (arsenm) [inline llvm/test/MachineVerifier/test_abd_su.mir:11]: Missing the error check?

Comment (tschuett) [reply llvm/test/MachineVerifier/test_abd_su.mir:11]: I removed the C++ error checking code under the assumption that the parameters are types and not ptypes and some other part of the system checks for misbehaviour.

Comment (arsenm) [reply llvm/test/MachineVerifier/test_abd_su.mir:11]: But that does not remove the need to check that the correct error is produced. I also see the case with a pointer type in each operand is not rejected

**Pull request merged**